_posts/2020-05-26-introduction-to-spring-webflux.md

```diff
@@ -0,0 +1,170 @@
+
+In our case we solved the issue by always returning a non-empty value, usually the same value which we pass to the function as a parameter:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .map(b -> operationXxx(b)));
+}
+
+Mono<Bar> checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return Mono.just(bar);
+  }
+  return Mono.error(new OperationDeniedException());
+}
+```
+
+This works as you want - `checkOperationXxx()` does the necessary checks and `operationXxx()` is executed only on a success case.
+
+#### Nothing happens before you subscribe
+
+With reactive code, you describe the steps of the operation, but nothing really gets executed before you call `subscribe()`. This is not that clear in WebFlux context, as Spring executes the `subscribe()` call to your WebFlux flow on your behalf, at least in case of REST Controller. You simply return the reactive result type from REST controller method and Spring will take care of the rest. In any case, it is good to understand.
+
+By the way, if you use WebClient from MVC side and wonder how to get the resulting object out from reactive context, just call `.block()`.
+
+#### Unit testing
+
+For unit testing reactive code I recommend using the official tool, which is the `reactor-test` library. Previously we used either `subscribe()`- or `block()`-approaches, but for some reason they were not always reliable. Better to stick with the recommended way. Reactor test library comes with `StepVerifier` class for unit testing reactive flows. Here's an example:
+```java
+@Test
+void testSuccessCase() {
+  StepVerifier.create(doSomething())
+    .expectNextMatches(f -> f.equals(expectedFoo))
+    .verifyComplete();
+}
+
+@Test
+void testErrorCase() {
+  StepVerifier.create(doSomething())
+    .expectErrorSatisfies(error -> {
+      // Assertions for error
+    })
+    .verify();
+}
+```
+
+For testing your reactive REST Controllers you can use Spring `WebTestClient`:
+```java
+@Test
+void testGetSomething() {
+
+  webTestClient.get()
+    .uri("/something")
+    .exchange()
+    .expectStatus().isOk()
+    .expectBody()
+    .jsonPath("$.value").isEqualTo(foo.getValue());
+}
+```
+
+How about testing your WebClient calls? Let's say you call some backend system to fetch some data and would like to test that functionality. While there's no built-in tool in Spring for such a case, you can use [MockWebServer](https://github.com/square/okhttp/tree/master/mockwebserver) to mock the backend. I believe Spring developers use it too for their testing purposes. Do note that you need to override both `MockWebServer` and `okhttp` library versions, because the versions coming in from Spring Boot dependency management are too old (at least in Spring Boot 2.3.0).
```

Comment:
'you call some backend system' -> 'you call a backend system'
'you can use [MockWebServer]' -> 'you can use the [MockWebServer]'
'I believe Spring developers use it too for their testing purposes.' -> 'I believe Spring developers use it for their testing purposes, too.'
'from Spring Boot' -> 'from the Spring Boot'